pymeasure/instruments/agilent/agilentE4407B.py

```diff
@@ -0,0 +1,637 @@
+        ":SENS:AVER:TYPE?",
+        ":SENS:AVER:TYPE %g",
+        """ A string property that represents the average type. LPOW is log power
+        and POW is linear power. This property can be set.
+        """,
+        validator=strict_discrete_set,
+        values=["LPOW", "POW"],
+    )
+
+    emi_detector_type = Instrument.control(
+        ":SENS:DET:EMI?",
+        ":SENS:DET:EMI %g",
+        """ A string property that represents the detector type.
+        This property can be set.
+        """,
+        validator=strict_discrete_set,
+        values=["QPE", "AVER", "OFF"],
+    )
+    emi_view_type = Instrument.control(
+        ":SENS:DET:EMI:VIEW?",
+        ":SENS:DET:EMI:VIEW %g",
+        """ A string property that represents the detector type.
+        This property can be set.
+        """,
+        validator=strict_discrete_set,
+        values=["POS", "EMI"],
+    )
+    qp_detector_gain = Instrument.control(
+        ":SENS:POW:QPG?",
+        ":SENS:POW:QPG %g",
+        """ Turn on or off the linear x10 gain stage in the quasi-peak and emi average detector only valid with the emi detector enabled.
+        This property can be set.
+        """,
+        validator=strict_discrete_set,
+        values=[0, 1, " ON", "OFF"],
+    )
+    input_attenuation = Instrument.control(
+        ":SENS:POW:ATT?",
+        ":SENS:POW:ATT %g",
+        """ A floating point property that represents the input attenuation
+        in dB. This property can be set.
+        """,
+        validator=truncated_discrete_set,
+        values=[0, 5, 10, 15, 20, 25, 30, 35, 40, 45, 50, 55, 60, 65],
+    )
+    input_attenuation_auto = Instrument.control(
+        ":SENS:POW:ATT:AUTO?",
+        ":SENS:POW:ATT:AUTO %g",
+        """ A boolean property that represents the input attenuation
+        auto mode. This property can be set.
+        """,
+    )
+    max_mixer_power = Instrument.control(
+        ":SENS:POW:MIX:RANG?",
+        ":SENS:POW:MIX:RANG %g",
+        """ A floating point property that represents the maximum mixer power
+        in dBm. This property can be set.
+        """,
+    )
+    # system commands
+
+    def reset(self):
+        """
+        A command that resets the instrument.
+        """
+        self.write("*RST")
+
+    def preset(self):
+        """Reset the instrument to the preset conditions."""
+        self.write("SYST:PRES")
+
+    def persistant_reset(self):
+        """Reset the instrument persistant state vakues to there factory defaults."""
+        self.write("SYST:PRES:PERS")
+
+    def save_user_preset(self):
+        """Save the current instrument state as a user preset."""
+        self.write("SYST:PRES:USER:SAVE")
+
+    def abort(self):
+        """
+        A command that aborts the sweep or measurement in progress.
+        """
+        self.write("ABOR")
+
+    def hardware_configuration(self):
+        """
+        A command that returns information about the current hardware in the instrument.
```

Comment:
Note that `write()` does not return anything, you'd have to do `return self.ask("SYS:CONF:HARD?")` to get the reply, too.
Same below.